intelmq/bots/outputs/kafka/output.py

```diff
@@ -0,0 +1,126 @@
+            self.avro_key_schema = avro.loads(
+                dumps(load_configuration(self.parameters.avro_key_schema_file)))
+
+            self.avro_topic_schema = load_configuration(self.parameters.avro_topic_schema)
+
+            self.avro_topic_field = getattr(
+                self.parameters, 'avro_topic_field', None)
+            self.avro_schema_registry = getattr(
+                self.parameters, 'avro_schema_registry', None)
+
+            # Build a list of producers for each destination topic
+            self.producer = AvroProducer({
+                                         'bootstrap.servers': self.broker_list,
+                                         'schema.registry.url': self.avro_schema_registry
+                                         },
+                                         default_key_schema=self.avro_key_schema,
+                                         default_value_schema=self.avro_value_schema
+                                         )
+        else:
+            self.producer = Producer({
+                                     'bootstrap.servers': self.broker_list
+                                     })
+
+        if isinstance(self.flatten_fields, str):
+            self.flatten_fields = self.flatten_fields.split(',')
+
+    def process(self):
+        event = self.receive_message()
+        event_dict = event.to_dict(hierarchical=False)
+
+        for field in self.flatten_fields:
+            if field in event_dict:
+                val = event_dict[field]
+                # if it's a string try to parse it as JSON
+                if isinstance(val, str):
+                    try:
+                        val = loads(val)
+                    except ValueError:
+                        pass
+                if isinstance(val, Mapping):
+                    for key, value in val.items():
+                        event_dict[field + '.' + key] = value
+                    event_dict.pop(field)
+
+        event_dict = replace_keys(event_dict)
+
+        if self.enable_avro is False:
+            self.producer.produce(self.kafka_topic, dumps(event_dict).encode('utf-8'), callback=self.delivery_report)
+            self.acknowledge_message()
+            self.kafka.flush()
+        else:
+            try:
+                key_field = str(event_dict[self.avro_topic_field]).replace('.', '_')
+            except KeyError:
+                self.logger.debug('Event %s has no field %s, dropping.', format(event_dict), format(self.avro_topic_field))
+                self.acknowledge_message()
+                return
+
+            submit_key = {'indicator': event_dict[key_field]}
+            event_topic = self.avro_topic_schema[key_field]
+            self.logger.debug('Shipped %s to topic: %s', format(submit_key), format(event_topic))
+            self.producer.produce(topic=event_topic, value=event_dict, key=submit_key)
+            self.acknowledge_message()
+            self.producer.poll(0)
```

Comment:
Better do this check before acknowledging and raise an error if that happened so the message does not get lost.

Reply:
It doesn't look like the AvroProducer has much in terms of error handling, looking at it now, i don't think that poll() does anything. I think i'm just going to have to catch serialization errors on produce.

[Here's a link](https://github.com/confluentinc/confluent-kafka-python/blob/fb9d973a69e9b86afe50dd4260c4f20850bef815/confluent_kafka/avro/__init__.py#L16)

Reply:
If a message could not be successfully delivered, `self.acknowledge()` should not be called. For error handling (and logging), an Exception should be raised.